pkg/services/ngalert/schedule/schedule.go

```diff
@@ -324,11 +335,22 @@ func (sch *schedule) ruleRoutine(grafanaCtx context.Context, key ngmodels.AlertR
 	evalDuration := sch.metrics.EvalDuration.WithLabelValues(orgID)
 	evalTotalFailures := sch.metrics.EvalFailures.WithLabelValues(orgID)
 
-	clearState := func() {
-		states := sch.stateManager.ResetStateByRuleUID(grafanaCtx, key)
-		expiredAlerts := FromAlertsStateToStoppedAlert(states, sch.appURL, sch.clock)
-		if len(expiredAlerts.PostableAlerts) > 0 {
-			sch.alertsSender.Send(key, expiredAlerts)
+	clearState := func(ctx context.Context, err error) {
+		// clean up the state only if the reason for stopping the evaluation loop is that the rule was deleted/paused
+		isRuleDeleted := errors.Is(err, errRuleDeleted)
+		isRulePaused := errors.Is(err, errRulePaused)
+		if isRuleDeleted || isRulePaused {
```

Comment:
There are 3 possible reasons to call this function: rule is deleted, rule is paused and rule is updated. Although 1st and 3rd are similar, semantically they are different. Let's make it expect nil err and declare it as update operation

Reply:
After thinking more, I think that conversion of error to reason should be isolated to handler of ctx.Done(). This method should accept the reason (state reason string). This along with changes in state manager to use single method for resetting state would reduce this diff.